java/ql/src/experimental/Security/CWE/CWE-200/AndroidWebResourceResponse.qll

```diff
@@ -0,0 +1,82 @@
+/** Provides Android methods relating to web resource response. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * The Android class `android.webkit.WebResourceRequest` for handling web requests.
+ */
+class WebResourceRequest extends RefType {
+  WebResourceRequest() { this.hasQualifiedName("android.webkit", "WebResourceRequest") }
+}
+
+/**
+ * The Android class `android.webkit.WebResourceResponse` for rendering web responses.
+ */
+class WebResourceResponse extends RefType {
+  WebResourceResponse() { this.hasQualifiedName("android.webkit", "WebResourceResponse") }
+}
+
+/** The `shouldInterceptRequest` method of Android's `WebViewClient` class. */
+class ShouldInterceptRequestMethod extends Method {
+  ShouldInterceptRequestMethod() {
+    this.hasName("shouldInterceptRequest") and
+    this.getDeclaringType().getASupertype*() instanceof TypeWebViewClient and
+    this.getReturnType() instanceof WebResourceResponse
+  }
+}
+
+/** A method call to `setWebViewClient` of `WebView`. */
+class SetWebViewClientMethodAccess extends MethodAccess {
+  SetWebViewClientMethodAccess() {
+    this.getMethod().hasName("setWebViewClient") and
+    this.getMethod().getDeclaringType().getASupertype*() instanceof TypeWebView and
+    this.getMethod().getParameterType(0) instanceof TypeWebViewClient
+  }
+}
+
+/** A sink representing a constructor call of `WebResourceResponse` in Android `WebViewClient`. */
+class WebResourceResponseSink extends DataFlow::Node {
+  WebResourceResponseSink() {
+    exists(ConstructorCall cc |
+      cc.getConstructedType() instanceof WebResourceResponse and
+      (
+        this.asExpr() = cc.getArgument(2) and cc.getNumArgument() = 3 // WebResourceResponse(String mimeType, String encoding, InputStream data)
+        or
+        this.asExpr() = cc.getArgument(5) and cc.getNumArgument() = 6 // WebResourceResponse(String mimeType, String encoding, int statusCode, String reasonPhrase, Map<String, String> responseHeaders, InputStream data)
+      )
+    )
+  }
+}
+
+/**
+ * Value step from a fetching url call of `WebView` to `WebViewClient`.
+ */
+private class FetchUrlStep extends AdditionalValueStep {
```

Comment:
I'm dubious about this step. `webview.loadUrl(tainted)` is already a sink for `java/android/unsafe-android-webview-fetch`, which means that everything that gets reported by this query using this step will also be reported by `java/android/unsafe-android-webview-fetch` already.

On the other hand, a user could run only one of the two queries, and it makes sense that both always report all their results, so this is probably fine _IF_ we believe there are reasonable results that don't go through the step (i.e. don't happen inside `shouldInterceptRequest`). Otherwise, this becomes just a specialization of the other query.

For instance, looking at the tests, it seems that all of the test cases would get picked up by `java/android/unsafe-android-webview-fetch`.